jadx-gui/src/main/java/jadx/gui/ui/DnDTabbedPane.java

```diff
@@ -0,0 +1,353 @@
+package jadx.gui.ui;
+
+import java.awt.*;
+import java.awt.dnd.DragSource;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.util.Objects;
+import java.util.Optional;
+
+import javax.swing.*;
+import javax.swing.plaf.metal.MetalTabbedPaneUI;
+
+class DnDTabbedPane extends JTabbedPane {
+	protected static final Rectangle RECT_LINE = new Rectangle();
+	private static final int SCROLL_SIZE = 20; // Test
+	private static final int BUTTON_SIZE = 30; // XXX 30 is magic number of scroll button size
+	private static final int LINE_SIZE = 3;
+	private static final Rectangle RECT_BACKWARD = new Rectangle();
+	private static final Rectangle RECT_FORWARD = new Rectangle();
+	// private final DropMode dropMode = DropMode.INSERT;
+	protected int dragTabIndex = -1;
+	private transient DnDTabbedPane.DropLocation dropLocation;
+
+	protected DnDTabbedPane() {
+		super();
+		Handler h = new Handler();
+		addMouseListener(h);
+		addMouseMotionListener(h);
+		addPropertyChangeListener(h);
+	}
+
+	public static boolean isTopBottomTabPlacement(int tabPlacement) {
+		return tabPlacement == SwingConstants.TOP || tabPlacement == SwingConstants.BOTTOM;
+	}
+
+	private void clickArrowButton(String actionKey) {
+		JButton scrollForwardButton = null;
+		JButton scrollBackwardButton = null;
+		for (Component c : getComponents()) {
+			if (c instanceof JButton) {
+				if (Objects.isNull(scrollForwardButton)) {
+					scrollForwardButton = (JButton) c;
+				} else if (Objects.isNull(scrollBackwardButton)) {
+					scrollBackwardButton = (JButton) c;
+				}
+			}
+		}
+		JButton button = "scrollTabsForwardAction".equals(actionKey) ? scrollForwardButton : scrollBackwardButton;
+		Optional.ofNullable(button).filter(JButton::isEnabled).ifPresent(JButton::doClick);
+
+		// // ArrayIndexOutOfBoundsException
+		// Optional.ofNullable(getActionMap())
+		// .map(am -> am.get(actionKey))
+		// .filter(Action::isEnabled)
+		// .ifPresent(a -> a.actionPerformed(new ActionEvent(this, ActionEvent.ACTION_PERFORMED, null, 0,
+		// 0)));
+		// // ActionMap map = getActionMap();
+		// // if (Objects.nonNull(map)) {
+		// // Action action = map.get(actionKey);
+		// // if (Objects.nonNull(action) && action.isEnabled()) {
+		// // action.actionPerformed(new ActionEvent(this, ActionEvent.ACTION_PERFORMED, null, 0, 0));
+		// // }
+		// // }
```

Comment:
Don't leave commented code.

Reply:
this is not my code, that's why I didn't clear it

Reply:
Hm, it is even worse, because now we need to add copyright and verify license of this code.

Reply:
Code from [here](https://java-swing-tips.blogspot.com/2010/02/tabtransferhandler.html?m=1)